docs/arcadia-users-group/20231128-testing-datasets/lesson.md

```diff
@@ -0,0 +1,84 @@
+The creation of each test data set was encoded in the test itself.
+
+Below we discuss strategies for creating test data sets to help maintain some of these desirable qualities in real code bases.
+
+## Test data for functions and scripts
+
+### Creating data within the test itself
+
+As seen in our previous lesson and described above, one strategy is to create test data within the test using built-in functions or data structures.
+When possible, this is a great strategy.
+These test data are small and self-contained and the correct answer to the test is easy to figure out.
+Functions like `[]` (list creation), or `range()` in Python and `c()`, `rep()`, `seq()`, `data.frame()`, and `list()` in R can be used to create test data that are relevant to a function.
+Some data-creation functions might be non-deterministic, but setting the seed ensures that the same output are produced each time.
+
+### Testing image outputs
+
+Often times, a function or a script produces an image output.
+Images are trickier to test than some other outputs, but can still be used in unit tests.
+If a function produces the image and an image object is produced in your coding environment, you can sometimes check the metadata of the image to make sure it matches expected values.
+This includes dimensions, color profiles, or file formats.
+These types of tests have the built-in benefit of testing whether the function produces an image successfully on top of confirming attributes about the image.
+
+Even if you check metadata attributes about an image, the image itself might still be different than expected.
+If you want to test the output image more thoroughly, you can compare the output produced by a test to a reference image stored in your test data.
+Both checksum comparisons and pixel comparisons are useful in this scenario.
+
+### Match md5sum of output files
+
+Checksums can also be used to compare the results of a function against other types of files.
+For example, if you have a function that produces a data frame, you can write that object to a CSV file.
+Then, you could compare the output CSV file against a reference CSV file by their checksums.
+When using this strategy, it might be necessary to sort your data frame prior to writing a file.
+
+## Test data sets for workflow integration tests
+
+Identifying a small data set that will quickly run through the entire workflow can be tricky.
+Below we cover some strategies to help achieve this.
+
+### Subsetting input data
+
+Sometimes, you can subset your data by running `head` or `tail` on your input files, or sub sampling input samples to a lower number, and your workflow will still run.
+This is a great option when it works!
+Often times, however, this strategy will fail as sub sampling in this manner will cause some tools in your workflow to produce no results, rendering the rest of the workflow unable to run or producing blank, meaningless outputs.
```

Comment:
````suggestion

For example, let's try taking just the first 1,000 lines of `/usr/share/dict/web2a` as a test file:

```
!head -1000 /usr/share/dict/web2a > /tmp/web2a.head
distribution_in_file('/tmp/web2a.head', '/tmp/web2a.head.txt')
```

Runtime is much faster. Unfortunately, results are significantly different - because the input file is sorted, we're testing words that are more likely to start with letters like 'a', skewing the character distribution. Examining `/tmp/web2a.txt` vs `/tmp/web2a.head.txt`, we see that the letter 'a' appears at a frequency of 6.7% in the former, but 12.1% in the latter.
Often times, simple subsetting strategies like this will fail in even more problematic ways -- subsampling in this manner can cause some tools in your workflow to produce no results, rendering the rest of the workflow unable to run or producing blank, meaningless outputs.

In this case can try sampling random lines from the file, instead of the first lines:

```
!sort -R /usr/share/dict/web2a | head -1000 > /tmp/web2a.rand
distribution_in_file('/tmp/web2a.rand', '/tmp/web2a.rand.txt')
```

The frequency of 'a' is again 6.7%, and the distribution as a whole more similar to the distribution over the full dataset.

````